### PR TITLE
fix(VTextField): use intersect to recalculate elements width

### LINKED ISSUE #9299: [Bug Report] Incorrect text field size calculation inside expansion panel
### Environment
**Vuetify Version:** 2.1.2
**Vue Version:** 2.6.10
**Browsers:** Chrome 77.0.3865.90
**OS:** Linux x86_64

### Steps to reproduce
1. Open expansion panel
2. Focus second field

### Expected Behavior
Correct notch width in border
Correct label positioning in second field

### Actual Behavior
Incorrect notch width in border
Incorrect label positioning in second field

### Reproduction Link
<a href="https://codepen.io/djaler/pen/oNNXywM" target="_blank">https://codepen.io/djaler/pen/oNNXywM</a>

### Other comments
Elements width calculated while they are hidden, so offsetWidth = 0

### PR DESCRIPTION
**Pull request by Djaler**
fixes #9299

## Description
Use IntersectionObserver (if it available) to recalculate element width after input becomes visible, 

## Motivation and Context
Cause they have zero width on mount in hidden state (inside expansion panel, for example)
https://github.com/vuetifyjs/vuetify/issues/9299

## How Has This Been Tested?
visually

## Markup:
https://codepen.io/djaler/pen/oNNXywM

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)
- [ ] I've added new examples to the kitchen (applies to new features and breaking changes in core library)

### CODE REVIEW

**Comment by Djaler** [comment]
@johnleider this is improved version of #9300

**Comment by johnleider** [comment]
@Djaler Could you message me in the community please, https://community.vuetifyjs.com

**Comment by Djaler** [comment]
@johnleider I updated a code according to what we discussed. Also, @sh7dm, I added a test to mixin

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        this.onObserve(entries)
+      })
+
+      observer.observe(this.$el)
+    },
+    methods: {
+      onObserve (entries: IntersectionObserverEntry[]) {
+        const isVisible = Boolean(entries.find(entry => entry.isIntersecting))
+
+        if (isVisible) {
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: reduce block nesting. If !visible can return early

**Comment by johnleider** [comment]
@KaelWD can you look over the mixin for inaccuracies in typescript implementation.

File: packages/vuetify/src/mixins/intersectable/index.ts (hunk 1 above)
Comment (Djaler) [reply packages/vuetify/src/mixins/intersectable/index.ts]: @johnleider done

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import Vue from 'vue'
+import { consoleWarn } from '../../util/console'
+
+export default function intersectable (options: { onVisible: string[] }) {
+  if (!('IntersectionObserver' in window)) {
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: this will break on SSR

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        this.onObserve(entries)
+      })
+
+      observer.observe(this.$el)
+    },
+    methods: {
+      onObserve (entries: IntersectionObserverEntry[]) {
+        const isVisible = Boolean(entries.find(entry => entry.isIntersecting))
+
+        if (!isVisible) {
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: this can be one lined

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+          return
+        }
+
+        for (let i = 0, length = options.onVisible.length; i < length; i++) {
+          const callback = (this as any)[options.onVisible[i]]
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: blank line between variable declaration and block logic

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+          if (typeof callback === 'function') {
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: opt for early return in this conditional.

```
if (typeof callback === 'function') {
  callback()

  return
}
```

Comment (Djaler) [reply packages/vuetify/src/mixins/intersectable/index.ts]: But we need to call all callbacks, not just first

File: packages/vuetify/src/mixins/intersectable/index.ts (hunk 2 above)
Comment (Djaler) [reply packages/vuetify/src/mixins/intersectable/index.ts]: Because of the `window`?

Comment (johnleider) [reply packages/vuetify/src/mixins/intersectable/index.ts]: You have to check for the existence of the window

File: packages/vuetify/src/mixins/intersectable/index.ts (hunk 5 above)
Comment (johnleider) [reply packages/vuetify/src/mixins/intersectable/index.ts]: sorry, `continue`

**Comment by Djaler** [comment]
@johnleider done

**Comment by Djaler** [comment]
@joeledenberg can you check it?

**Comment by Djaler** [comment]
@johnleider can you check it?

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    mounted () {
+      const observer = new IntersectionObserver((
+        entries: IntersectionObserverEntry[] = [],
+      ) => {
+        this.onObserve(entries)
+      })
+
+      observer.observe(this.$el)
+    },
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: add line between mounted and methods

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    })
+  }
+
+  return Vue.extend({
+    name: 'intersectable',
+
+    mounted () {
+      const observer = new IntersectionObserver((
+        entries: IntersectionObserverEntry[] = [],
+      ) => {
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: change to `) => this.onObserve(entries)`

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import Vue from 'vue'
+import { consoleWarn } from '../../util/console'
+
+export default function intersectable (options: { onVisible: string[] }) {
+  if (typeof window === 'undefined' || !('IntersectionObserver' in window)) {
+    return Vue.extend({
+      name: 'intersectable',
+      // do nothing because intersection observer is not available
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: Move to line 6

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import Vue from 'vue'
+import { consoleWarn } from '../../util/console'
+
+export default function intersectable (options: { onVisible: string[] }) {
+  if (typeof window === 'undefined' || !('IntersectionObserver' in window)) {
+    return Vue.extend({
+      name: 'intersectable',
```
Comment (johnleider) [inline packages/vuetify/src/mixins/intersectable/index.ts]: one line this, `return Vue.extend({ name: intersectable })`

**Comment by Djaler** [comment]
How can I add directive from mixin?

**Comment by Djaler** [comment]
@johnleider how can we add directive from mixin?

**Comment by KaelWD** [comment]
> @KaelWD can you look over the mixin for inaccuracies in typescript implementation.

It's alright. Could make the options type safe by being a wrapping function instead of a mixin factory, but that gets pretty complicated.

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+      // do nothing because intersection observer is not available
+    })
+  }
+
+  return Vue.extend({
+    name: 'intersectable',
+
+    mounted () {
+      const observer = new IntersectionObserver((
+        entries: IntersectionObserverEntry[] = [],
```
Comment (KaelWD) [inline packages/vuetify/src/mixins/intersectable/index.ts]: Don't need this, the type is inferred and correct me if I'm mistaken but isn't `entries` always defined?

File: packages/vuetify/src/mixins/intersectable/index.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+      name: 'intersectable',
+      // do nothing because intersection observer is not available
+    })
+  }
+
+  return Vue.extend({
+    name: 'intersectable',
+
+    mounted () {
+      const observer = new IntersectionObserver((
```
Comment (KaelWD) [inline packages/vuetify/src/mixins/intersectable/index.ts]: Do we need to tear this down when the component is destroyed?

**Comment by Djaler** [comment]
> The mixin could have everything else though and the implementing component would just have to make sure to use the directive somewhere.

Do you mean, in implementing component we will have something like 
```
directives: [
          {
            name: 'intersect',
            value: {
              handler: this.onObserve,
            },
          },
        ],
```, where `onObserve` provided by mixin?

**Comment by KaelWD** [comment]
You can't do that, implementing component would have https://github.com/vuetifyjs/vuetify/pull/9300/files#diff-dc0c0fccb5e57dfaca5093b2096562fdR386-R393 in its render function somewhere. See my edit for another idea.

**Comment by Djaler** [comment]
> ```js
> mounted () {
>   Intersect.inserted(this.$el, options)
> }
> destroyed () {
>   Intersect.unbind(this.$el)
> }
> ```

Do you mean to get rid of mixin at all?

**Comment by KaelWD** [comment]
No, the mixin can do that.

**Comment by Djaler** [comment]
Oh, I did not get it that Intersection.inserted is a way to initialize directive. Yeah, this could work

**Comment by Djaler** [comment]
@KaelWD @johnleider done

**Comment by Djaler** [comment]
So what about it, guys?

**Pull request merged**